src/fn/upload-custom-translations.js

```diff
@@ -81,6 +114,76 @@ function parse(filePath, options) {
   });
 }
 
+function checkTranslations(translations, lang, templatePlaceholders) {
+  let mf = null;
+  try {
+    mf = new MessageFormat(lang);
+  } catch (e) {
+    log.warn(`Cannot check '${lang}' translations: ${e.message}`);
+  }
+  const placeholders = extractPlaceholdersFromTranslations(translations);
+  let formatErrorsFound = 0,  placeholderErrorsFound = 0, emptiesFound = 0;
+  for (const [msgKey, msgSrc] of Object.entries(translations)) {
+    if (!msgSrc) {
+      emptiesFound++;
+    } else if (/{{[\s\w.#^/'|]+}}/.test(msgSrc)) {
+      if (templatePlaceholders) {
+        const placeholder = placeholders[msgKey];
+        if (placeholder) {
+          const templatePlaceholder = templatePlaceholders[msgKey];
+          if (!templatePlaceholder) {
+            log.error(`Cannot compile '${lang}' translation with key '${msgKey}' ` +
+              `has placeholders, but the key does not match any of ${EN_FILE}\n` +
+              'You can use messages-ex.properties to add placeholders missing from the reference context.');
+            placeholderErrorsFound++;
+          } else {
+            const foundAllPlaceholders = placeholder.every(el => templatePlaceholder.includes(el));
+            if (!foundAllPlaceholders) {
+              log.error(`Cannot compile '${lang}' translation with key '${msgKey}' ` +
+                `has placeholders that do not match those of ${EN_FILE}\n` +
+                'You can use messages-ex.properties to add placeholders missing from the reference context.');
+              placeholderErrorsFound++;
+            }
+          }
+        }
+      }
+    } else if (mf && typeof msgSrc === 'string') {
+      try {
+        mf.compile(msgSrc);
+      } catch (e) {
+        log.error(`Cannot compile '${lang}' translation ${msgKey} = '${msgSrc}' : ${e.message}`);
+        formatErrorsFound++;
+      }
+    }
+  }
+  if (emptiesFound > 0) {
+    log.warn(transEmptyMsg({EMPTIES: emptiesFound}) + ` '${lang}' translations`);
+  }
+  if (formatErrorsFound > 0 || placeholderErrorsFound > 0) {
+    log.error(transErrorsMsg({ERRORS: formatErrorsFound + placeholderErrorsFound})
+              + ` '${lang}' translations`);
+    process.exit(-1);
+  }
+}
+
+function extractPlaceholdersFromTranslations(translations, extraPlaceholders = {}) {
+  // Extract from github.com/medic/cht-core/blob/master/scripts/poe/lib/utils.js
+  const result = {};
+  for (const [msgKey, msgSrc] of Object.entries(translations)) {
+    let placeholders = typeof msgSrc === 'string' ? msgSrc.match(/{{[\s\w.#^/'|]+}}/g) : null;
```

Comment:
This regex is identical to the one above, so when you pull that out as a constant you can use it here too.

Reply:
Yes note that this one has the `/g` suffix at the end to make it "global" while the other not, but thinking about it again I can use the global expression there too.

Reply:
Done.

Reply:
@garethbowen I had to rollback the regex constant refactor, turns out regex/g are mutable objects :facepalm: : 

```
> // Not using constant the result is predictable:
undefined
> /{{[\s\w.#^/'|]+}}/g.test('Hey {{you}}!')
true
> /{{[\s\w.#^/'|]+}}/g.test('Hey {{you}}!')
true
> // Now using a constant the result change on each execution:
undefined
> const MUSTACHE_MATCHER = /{{[\s\w.#^/'|]+}}/g
undefined
> MUSTACHE_MATCHER.test('Hey {{you}}!')
true
> MUSTACHE_MATCHER.test('Hey {{you}}!')
false
> MUSTACHE_MATCHER.test('Hey {{you}}!')
true
```

Replacing the `/g` at the end of the regex by just `/` solve the problem with the `test(...)` method, but is needed by the `match(...)` method used in another part of the code. I can also remove the `/g` and use `matchAll` instead, but you need Node 12+.

Reply:
Wow, nice find!

I think you can use `match` and check for `null` instead of using `test`, eg:

```
> 'Hey {{you}}!'.match(MUSTACHE_MATCHER);
["{{you}}"]
> 'Hey '.match(MUSTACHE_MATCHER);
null
```

Reply:
Moved again as constant but avoiding to use the `test()` method that mutes the regex constant state